File: assets/Scripts/player/controllers/PlayerGameplayController.js
// FILE: Scripts/player/controllers/PlayerGameplayController.js
"use strict";

const MovementSystem = require("../systems/MovementSystem.js");
const ShootSystem = require("../systems/ShootSystem.js");

function req(v, msg) {
    if (v == null) throw new Error(msg);
    return v;
}

/**
 * PlayerGameplayController
 *
 * Owns gameplay loop for PlayerPawn:
 *  - beginFrame(dt)
 *  - syncPose()
 *  - movement.update(frame, characterCfg)
 *  - shoot.update(frame, ownerBodyId)
 */
class PlayerGameplayController {
    constructor(player) {
        this.player = req(player, "[PlayerGameplayController] player is required");
        this.movement = null;
        this.shoot = null;
    }

    onStart() {
        const pawn = req(this.player.pawn, "[PlayerGameplay] player.pawn is required");

        if (typeof pawn.beginFrame !== "function") throw new Error("[PlayerGameplay] pawn.beginFrame required");
        if (typeof pawn.syncPose !== "function") throw new Error("[PlayerGameplay] pawn.syncPose required");
        if (!pawn.frame) throw new Error("[PlayerGameplay] pawn.frame is required");
        if ((pawn.bodyId | 0) <= 0) throw new Error("[PlayerGameplay] pawn.bodyId must be > 0");

        this.movement = new MovementSystem((pawn.cfg && pawn.cfg.movement) || null);
        this.shoot = new ShootSystem(pawn);
    }

    onUpdate(dt) {
        const pawn = req(this.player.pawn, "[PlayerGameplay] player.pawn is required");

        pawn.beginFrame(dt);
        pawn.syncPose();

        const frame = pawn.frame;
        if (!frame.input || typeof frame.input !== "object") {
            throw new Error("[PlayerGameplay] frame.input must be an object (InputRouter state)");
        }

        this.movement.update(frame, pawn.characterCfg);
        this.shoot.update(frame, pawn.bodyId | 0);
    }

    onStop() {
        if (this.shoot && typeof this.shoot.destroy === "function") {
            try { this.shoot.destroy(); } catch (_) {}
        }
        this.shoot = null;
        this.movement = null;
    }
}

module.exports = { PlayerGameplayController };

File: assets/Scripts/player/controllers/PlayerGameplayController.lua
local M = {}
local luaRuntime = require("@builtin/lua_runtime")
local LuaClass = luaRuntime.LuaClass
local Error = luaRuntime.Error
local LuaConstruct = luaRuntime.LuaConstruct
MovementSystem = require("../systems/MovementSystem.lua")
ShootSystem = require("../systems/ShootSystem.lua")
PlayerGameplayController = LuaClass()
PlayerGameplayController.name = "PlayerGameplayController"
function PlayerGameplayController.prototype.lua_constructor(self)
    self.ctx = nil
    self.entity = nil
    self.movement = nil
    self.shoot = nil
end
function PlayerGameplayController.prototype.bind(self, ctx, entity)
    self.ctx = ctx
    self.entity = entity
    return self
end
function PlayerGameplayController.prototype.onStart(self)
    local pawn = self.entity
    if not pawn or KTypeOf(pawn.beginFrame) ~= "function" then
        error(
            LuaConstruct(Error, "[PlayerGameplay] entity must be PlayerPawn (beginFrame)"),
            0
        )
    end
    if KTypeOf(pawn.syncPose) ~= "function" then
        error(
            LuaConstruct(Error, "[PlayerGameplay] entity must be PlayerPawn (syncPose)"),
            0
        )
    end
    if not pawn.frame then
        error(
            LuaConstruct(Error, "[PlayerGameplay] pawn.frame is required"),
            0
        )
    end
    if bit32.bor(pawn.bodyId, 0) <= 0 then
        error(
            LuaConstruct(Error, "[PlayerGameplay] pawn.bodyId must be > 0"),
            0
        )
    end
    self.movement = LuaConstruct(MovementSystem, pawn.cfg and pawn.cfg.movement or nil)
    self.shoot = LuaConstruct(ShootSystem, pawn)
end
function PlayerGameplayController.prototype.onUpdate(self, dt)
    local pawn = self.entity
    pawn:beginFrame(dt)
    pawn:syncPose()
    local frame = pawn.frame
    if not frame.input or KTypeOf(frame.input) ~= "table" then
        error(
            LuaConstruct(Error, "[PlayerGameplay] frame.input must be an object (InputRouter state)"),
            0
        )
    end
    self.movement:update(frame, pawn.characterCfg)
    self.shoot:update(
        frame,
        bit32.bor(pawn.bodyId, 0)
    )
end
function PlayerGameplayController.prototype.onStop(self)
    if self.shoot and KTypeOf(self.shoot.destroy) == "function" then
        self.shoot:destroy()
    end
    self.shoot = nil
    self.movement = nil
end
M = {PlayerGameplayController = PlayerGameplayController}

return M
